src/rust/engine/ui/src/console_ui.rs

```diff
@@ -198,40 +109,299 @@ impl ConsoleUI {
+            // This is confirmed before creating a UI.
+            output_is_terminal: true,
+            hide_cursor: true,
+            // TODO: Adjust this.
+            frames_per_second: ConsoleUI::render_rate_hz() as f32,
+            throughput: false,
+            terminal_dimensions,
+            ..Default::default()
+          },
+        );
+
+        // Spawn a task to propagate stderr, which will exit automatically when the channel closes.
+        // TODO: There is a shutdown race here, where if the UI is torn down before exclusive access is
+        // dropped, we might drop stderr on the floor. That likely causes:
+        //   https://github.com/pantsbuild/pants/issues/13276
+        let _stderr_task = executor.spawn_blocking({
+          let mut tree = tree.clone();
+          move || {
+            while let Ok(stderr) = stderr_receiver.recv() {
+              tree.message_raw(stderr);
+            }
+          }
+        });
+
+        Ok(Instance::Prodash(ProdashInstance {
+          tasks_to_display: HashMap::new(),
+          tree,
+          handle,
+          terminal_dimensions,
+          executor,
+        }))
+      },
+      _ => {assert!(false); Err(String::from("Bad renderer option!"))},
+    };
+  }
+
+  pub fn render(&mut self, heavy_hitters: &HashMap<SpanId, (String, SystemTime)>) {
+    match self {
+      Instance::Indicatif(indicatif) => {
+        let tasks_to_display = &mut indicatif.tasks_to_display;
+
+        // Insert every one in the set of tasks to display.
+        // For tasks already here, the durations are overwritten.
+        tasks_to_display.extend(heavy_hitters.clone().into_iter());
+
+        // And remove the tasks that no longer should be there.
+        for (task, _) in tasks_to_display.clone().into_iter() {
+          if !heavy_hitters.contains_key(&task) {
+            tasks_to_display.swap_remove(&task);
+          }
+        }
+
+        let now = SystemTime::now();
+        for (n, pbar) in indicatif.bars.iter().enumerate() {
+          let maybe_label = tasks_to_display
+            .get_index(n)
+            .map(|(_, (label, start_time))| {
+              let duration_label = match now.duration_since(*start_time).ok() {
+                None => "(Waiting)".to_string(),
+                Some(duration) => format_workunit_duration_ms!((duration).as_millis()).to_string()
+              };
+              format!("{} {}", duration_label, label)
+            });
+
+          match maybe_label {
+            Some(label) => pbar.set_message(label),
+            None => pbar.set_message(""),
+          }
+        }
+      },
+      Instance::Prodash(prodash) => {
+        let tasks_to_display = &mut prodash.tasks_to_display;
+
+        // Finish any items that are no longer relevant.
+        for span_id in tasks_to_display.keys().cloned().collect::<Vec<_>>() {
+          if heavy_hitters.contains_key(&span_id) {
+            // NB `inc_by` moves the "worms" to help show continual progress
```

Comment:
Didn't know what to call the lengths of "=". So I settled on "worms", because, uh, that's what they look like.

Reply:
Heh.

The comment is stale I think: now `inc` rather than `inc_by`.